### PR TITLE
[electrophysiology_browser] HED Tag Support

### LINKED ISSUE #8910: [EEG] HED Tag Support
HED tag support, which is present in EEGNet, can be brought in to LORIS. This adds the ability to add and remove HED tags from event instances and attribute mappings (events.json).

This includes SQL patches and significant Loris-MRI changes.

This also includes the removal of "Annotations" in favour for "Events", in order to keep up with the latest BIDS specifications, and a CouchDB import.

### PR DESCRIPTION
**Pull request by jeffersoncasimir**
This PR adds more robust HED tag support. Closes #8910.
Depends on #9032 (should be merged before reviewing)
Loris-MRI counterpart: [#1041](https://github.com/aces/Loris-MRI/pull/1041)

It adds SQL tables to store HED schemas as hierarchical nodes and to store the assembled HED strings as node references using a modified linked list data structure, allowing for complex groupings.

Instance-level tags which were ingested in the `HED` column of `events.tsv` files can be viewed in the `EventManager` component and can be edited in a limited (see below) way in the `AnnotationForm` component.

There is a new component called `DatasetTagger` which is accessible via the "Dataset Tag Manager" button at the bottom left corner of the browser. This component enables the modification of HED tags at the dataset-wide level (`events.json`). This will associate HED tags to Event attributes for all events in the dataset containing the respective attributes. This component enables virtually unlimited groupings of HED tags, as well their addition and removal.

Limitations:
- "Value" HED tags are not supported
- "Expanded Definitions" are not supported
- Editing the "Level description" via the "Dataset Tag Manager" is not currently supported (coming soon)
- Session-level events.json overriding and assignment not currently supported.
- Adding HED tags on an event level is currently limited to: SCORE artifacts (if SCORE schema ingested) and tags that were previously added via the ingestion of the `events.json` or manually via the Dataset Tag Manager.

~TODO: Raisinbread modifications + HED tag inserts~

**MAKE SURE YOU ADD BOTH SCHEMAS WITH THE TOOL FIRST (or use RB):** [HED8.2.0.xml](https://zenodo.org/records/7877308/files/HED8.2.0.xml) [HED_score_1.1.0.xml](https://raw.githubusercontent.com/hed-standard/hed-schemas/main/library_schemas/score/hedxml/HED_score_1.0.0.xml)

### CODE REVIEW

**Comment by driusan** [comment]
Blocked by #9032

**Comment by laemtl** [comment]
@jeffersoncasimir Can you rebase?

**Review by laemtl — APPROVED** [review]

File: modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/DatasetTagger.tsx
Comment (driusan) [inline modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/DatasetTagger.tsx:607]: There should no be any console.logs. They are used for debugging and flagged by `make checkstatic`.

This looks like it's an error not a log.

**Review by driusan — CHANGES_REQUESTED** [review]

**Comment by laemtl** [comment]
@jeffersoncasimir Can you rebase?

**Comment by christinerogers** [comment]
added `blocking` because this is needed for release testing and for https://github.com/aces/Loris-MRI/pull/1041

File: modules/electrophysiology_browser/php/models/datasettags.class.inc
Comment (driusan) [inline modules/electrophysiology_browser/php/models/datasettags.class.inc:472]: This is not a useful error message to show up in the apache error log. No one reading it will know what it's talking about or even what module it's related to.

If it's necessary to log something here it should also be using a PSR logger so that the log level can be changed.

File: modules/electrophysiology_browser/php/models/datasettags.class.inc
Comment (driusan) [inline modules/electrophysiology_browser/php/models/datasettags.class.inc:609]: I'm not sure of the context of what the format of tag_string is but this seems like the type of function you might want to write unit tests for.

File: modules/electrophysiology_browser/php/models/electrophysioevents.class.inc
Comment (driusan) [inline modules/electrophysiology_browser/php/models/electrophysioevents.class.inc:324]: I need to review this part much more closely but I'm submitting the rest of the review for now.

**Review by driusan — CHANGES_REQUESTED** [review]

File: modules/electrophysiology_browser/jsx/react-series-data-viewer/src/eeglab/EEGLabSeriesProvider.tsx
Comment (jeffersoncasimir) [reply modules/electrophysiology_browser/jsx/react-series-data-viewer/src/eeglab/EEGLabSeriesProvider.tsx:299]: This is a placeholder for a self-hosted link to the [HED logo](https://www.hed-resources.org/en/latest/_images/croppedWideLogo.png). Can we get it hosted on images.loris.ca?

Comment (driusan) [reply modules/electrophysiology_browser/jsx/react-series-data-viewer/src/eeglab/EEGLabSeriesProvider.tsx:299]: Yes but I don't know how, you'll have to ask @laemtl

File: modules/electrophysiology_browser/php/models/electrophysioevents.class.inc
Comment (driusan) [inline modules/electrophysiology_browser/php/models/electrophysioevents.class.inc:317]: The file I/O here is not done safely. I haven't determined where the filePath comes from, but if it comes from the user (or ever might) this is very insecure. A user can craft a path with an arbitrary number of `../../` to overwrite any file on the filesystem that the web server has write access to. There is also no check around the `fopen` call below for the `''` case.

Luckily the fix is fairly simple. Use `\Utility::joinPath($dataDir, \Utility::resolvePath(filepath))` in all the places that calculate a path. This will resolve all the relative components of filePath before joining it to dataDir so that the path calculated can't escape $dataDir (and add an extra check for `''` or `$dataDir` in the if statements around the fopen calls.)

**Review by driusan — DISMISSED** [review]

**Comment by christinerogers** [comment]
basically co-dependent with https://github.com/aces/Loris-MRI/pull/1041 (blocked and blocking)

File: modules/electrophysiology_browser/jsx/react-series-data-viewer/src/eeglab/EEGLabSeriesProvider.tsx
Comment (jeffersoncasimir) [reply modules/electrophysiology_browser/jsx/react-series-data-viewer/src/eeglab/EEGLabSeriesProvider.tsx]: Resolved by 5229dcb8d62bbe24a6755fcaaec807783da6d781
The small logo has been added htdocs/images.

File: modules/electrophysiology_browser/php/models/datasettags.class.inc
Comment (jeffersoncasimir) [reply modules/electrophysiology_browser/php/models/datasettags.class.inc]: Resolved by 093b3f2e88f05d53b35c1c57a9c396c27bd08fac.
Removed

File: modules/electrophysiology_browser/php/models/datasettags.class.inc
Comment (jeffersoncasimir) [reply modules/electrophysiology_browser/php/models/datasettags.class.inc]: Resolved by 91a08a0ebbf267b8934c7f797f7d81964be6947a.

Logging now uses PSR\Logger

File: modules/electrophysiology_browser/php/models/datasettags.class.inc
Comment (jeffersoncasimir) [reply modules/electrophysiology_browser/php/models/datasettags.class.inc:633]: Good point. I opened an issue about writing unit tests for this and a few other functions in the near future: #9123.

File: modules/electrophysiology_browser/php/models/electrophysioevents.class.inc
Comment (jeffersoncasimir) [reply modules/electrophysiology_browser/php/models/electrophysioevents.class.inc]: Resolved by fcb9fe5efd802426b34de279723d4061543e196d

Using Utility functions

File: modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/Form.js
Comment (jeffersoncasimir) [reply modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/Form.js]: Resolved by c442f0e91e2787faff9b112ae6b62b4cdd01e2a5.

I did not create the file but I suspect it was created to bypass some style constraints, as I initially did with the `CheckboxElement` I added. For the other elements I added below it, it was because I was encountering an issue with being able to see the `SelectDropdown` component in `jsx/MultiSelectDropdown.js`. Turns out the there are some missing `export` keywords in  `jsx/MultiSelectDropdown.js` and I added them in the commit stated above.

I made some style changes to be able to remove the redundant components that I added. It now uses the existing Loris components.

**Pull request merged**
